doc: Document Platform Support Tiers (#245368)
NixOS/nixpkgs: doc/stdenv/platform-support-tiers.chapter.md

# Platform Support Tiers {#platform-support-tiers}

In the context of Nixpkgs, a *platform* can mean a CPU architecture, a
[cross-compilation](#chap-cross) target or a C library option.

Nixpkgs defines 7 platform *support tiers*, meaning levels of support for each platform.

**TODO**: change build successfully to a definition of "work".

## Tier 1 {#platform-support-tier-1}

These platforms receive the highest level of support, meaning that every
platform-compatible package in Nixpkgs must successfully build in the Hydra and OfBorg CIs.
**TODO**: this seems to not be entirely true for Hydra, but true for OfBorg.

> - `x86_64-linux`, `gcc` + `glibc`

Author: Should these be tables?

Reviewer: If we start with «copy what needs to be copied», let's keep RFC formatting. Then reformatting different parts of the text can be done at its own pace

## Tier 2 {#platform-support-tier-2}

Many of the packages are required to build successfully on these platforms in CI, the rest are supported on a best-effort basis by dedicated platform maintainers.

> **TODO**: link teams

Author: Problem is that the `@NixOS/aarch64-maintainers` mandated by the RFC was never created

- `aarch64-linux`, `gcc`+`glibc`, cross-compilation
- `mipsel-linux`, `gcc`+`glibc`
- `x86_64-linux`, `gcc`+`musl`

## Tier 4 {#platform-support-tier-4}

Some of the packages for these platforms are expected to work. The
[`exotic-platform-maintainers`
Team](https://github.com/orgs/NixOS/teams/exotic-platform-maintainers) is
responsible for this platform.

- `aarch64-none`
- `avr`
- `arm-none`
- `i686-none`
- `x86_64-none`
- `powerpc-none`
- `powerpcle-none`
- `x86_64-mingw32`
- `i686-mingw32`
- `x86_64-linux`, `gcc`+`musl` — static
- `x86_64-linux`, `clang`+`glibc`
- `x86_64-linux`, `clang`+`glibc` — `llvm` linker
- `x86_64-linux` — Android
- `aarch64-linux` — Android
- `armv{7,8}-linux` — Android

## Tier 5 {#platform-support-tier-5}

A small number of packages might build successfully on these platforms.

- `x86_64-linux`, `gcc`+`glibc` — static
- `x86_64-linux`, `gcc`+`glibc` — `llvm` linker

## Tier 6 {#platform-support-tier-6}

These platforms are defined in Nixpkgs, but no packages are expected to build on them.

- `wasm-wasi`
- `powerpc64le-linux`, `gcc`+`glibc`

## Tier 7 {#platform-support-tier-7}

No current support, but previous support or clear path to add support.

> - `aarch64-darwin`

Author: This seems to be false now, since OfBorg builds for this platform.

Reviewer: Yes, it looks like right now after all the work done, two macOS versions are mostly on par. And upstream pretends that the latter is the continuation of the former in some sense.

Reviewer 2: > I'm inclined towards leaving this at the state it was when the RFC was accepted, since I don't have the resources to update the list on my own.

In that case you should probably close this PR.

Adding inaccurate data to the documentation is not helpful.

Author: Well, it's inaccurate until someone fixes it, and seeing how nothing has happened in 3 years, this PR might at least get some things moving.

Author: But let's see if we gather more up to date info with this PR. Thank you for your contributions to that end.

> # Platform Support Tiers {#platform-support-tiers}

Reviewer 2: ```suggestion
# Non-normative Platform Support Tiers {#platform-support-tiers}
```

The RFC specifies these as "non-normative" (and also nearly *four years out of date* at this point).

> In the context of Nixpkgs, a *platform* can mean a CPU architecture, a
> [cross-compilation](#chap-cross) target or a C library option.

Reviewer 2: ```suggestion
In the context of Nixpkgs, a *platform* means an attrset which typechecks against `lib.systems.parsed.types.parsedPlatform`.  This includes a CPU architecture, bitwidth, and endianness and may also include: an OS kernel, environment (also called "C library"), and other data.
```

Nixpkgs defines 7 platform *support tiers*, meaning levels of support for each platform.

- **TODO**: change "build successfully" to a definition of "work" that includes tests?
- **TODO**: mention binary cache?

## Tier 1 {#platform-support-tier-1}

These platforms receive the highest level of support, meaning that every
platform-compatible package in Nixpkgs must successfully build in the Hydra and OfBorg CIs.
**TODO**: this seems to not be entirely true for Hydra, but true for OfBorg.

> - `x86_64-linux`, `gcc` + `glibc`

Reviewer 2: ```suggestion
- `x86_64-linux-gnu` using `gcc`
```

## Tier 2 {#platform-support-tier-2}

Many of the packages are required to build successfully on these platforms in CI, the rest are supported on a best-effort basis by dedicated platform maintainers.
**TODO**: link teams

- `aarch64-linux`, `gcc+glibc`
- `x86_64-darwin`, `clang+Darwin/macOS`

## Tier 3 {#platform-support-tier-3}

None of the packages for these platforms are required to build successfully in
CI, but it is expected that most of them will build successfully.

> - `i686-linux`, `gcc`+`glibc`
> - `armv{6,7,8}*-linux`, `gcc`+`glibc`
> - `armv{6,7,8}*-linux`, `gcc`+`glibc`, cross-compilation
> - `aarch64-linux`, `gcc`+`glibc`, cross-compilation
> - `mipsel-linux`, `gcc`+`glibc`
> - `x86_64-linux`, `gcc`+`musl`

Reviewer 2: ```suggestion
- `i686-linux-gnu`, `gcc`
- `armv{6,7,8}*-linux-gnu`, `gcc`
- `armv{6,7,8}*-linux-gnu`, `gcc`, cross-compilation
- `aarch64-linux-gnu`, `gcc`, cross-compilation
- `mipsel-linux-gnu`, `gcc`
- `x86_64-linux-musl`, `gcc`
```

## Tier 4 {#platform-support-tier-4}

Some of the packages for these platforms are expected to work. The
[`exotic-platform-maintainers`
Team](https://github.com/orgs/NixOS/teams/exotic-platform-maintainers) is
responsible for this platform.

- `aarch64-none`
- `avr`
- `arm-none`
- `i686-none`
- `x86_64-none`

> - `powerpc-none`
> - `powerpcle-none`

Reviewer 2: ```suggestion
- `powerpc-none`
- `powerpcle-none`
- `powerpc64le-linux-gnu`
- `mips64el-linux-gnuabi64`
```

All elements of the NixOS installer currently build on `powerpc64le-linux-gnu` and `mips64el-linux-gnuabi64`.

For `mips64el`, the PR to allow building the installer image itself has not yet merged, but all of its dependencies have:

- https://github.com/NixOS/nixpkgs/pull/194153

- `x86_64-mingw32`
- `i686-mingw32`

> - `x86_64-linux`, `gcc`+`musl` — static
> - `x86_64-linux`, `clang`+`glibc`
> - `x86_64-linux`, `clang`+`glibc` — `llvm` linker

Reviewer 2: ```suggestion
- `x86_64-linux-musl`, `gcc` — static
- `x86_64-linux-gnu`, `clang`
- `x86_64-linux-gnu`, `clang` — `llvm` linker
```

> - `x86_64-linux` — Android
> - `aarch64-linux` — Android
> - `armv{7,8}-linux` — Android

Reviewer 2: ```suggestion
- `x86_64-linux-android`
- `aarch64-linux-android`
- `armv{7,8}-linux-android`
```

## Tier 5 {#platform-support-tier-5}

A small number of packages might build successfully on these platforms.

> - `x86_64-linux`, `gcc`+`glibc` — static

Reviewer 2: I'm not sure this is correct.  I don't think static linking is truly possible with `glibc`.

Author: @7c6f434c what do you say?

Reviewer: We do have something that claims to be static linking with glibc. I am not sure which of the, say, NSS things it ends up doing with dynamic linking, but it does seem to link statically some stuff? As there are packages that depend on `glibc.static`, I tried to describe this build configuration as a platform…

> - `x86_64-linux`, `gcc`+`glibc` — `llvm` linker

Reviewer 2: ```suggestion
- `x86_64-linux-gnu`, `gcc` `llvm` linker
```

## Tier 6 {#platform-support-tier-6}

These platforms are defined in Nixpkgs, but no packages are expected to build on them.

- `wasm-wasi`

> - `powerpc64le-linux`, `gcc`+`glibc`

Reviewer 2: ```suggestion
```

The *entire NixOS installer* builds on `powerpc64le-linux`, and has built for quite some time:

- https://github.com/NixOS/nixpkgs/pull/192672

## Tier 7 {#platform-support-tier-7}

No current support, but previous support or clear path to add support.

- `aarch64-darwin`
- `i686-darwin`
- `x86_64-freebsd`
- `i686-solaris`
- `x86_64-illumos`

## Further reading {#platform-support-tiers-further-reading}

> For a formal definition of platform support tiers, see [RFC046](https://github.com/7c6f434c/rfcs/blob/platform-support-tiers/rfcs/0046-platform-support-tiers.md).

Author: What definition would you add here then?

> For a formal definition of platform support tiers, see [RFC046](https://github.com/7c6f434c/rfcs/blob/platform-support-tiers/rfcs/0046-platform-support-tiers.md).

Reviewer: I would stupidly add lines 55 to 275 from the RFC. This definitely covers «definition of support tiers» (and at least a half of it is definitely necessary). If this PR is «copy what RFC calls to copy» and updates (including formatting) are the next step, I think this is a reasonable way to handle definitions.
